### PR TITLE
Add isImageAvailable method to Metadata

### PR DESCRIPTION
**Pull request by jo66**
## Subject

This PR will allow use of default mosaic background (see [admin bundle PR #5366](https://github.com/sonata-project/SonataAdminBundle/pull/5366))

I am targeting this branch, because it's feature.

## Changelog

```markdown
### Added
- Added `isImageAvailable` method to `Metadata` class to allow use of default mosaic background
```

## To do
    
- [X] Update the tests

### CODE REVIEW

File: src/Meta/MetadataInterface.php [hunk 1]
```diff
@@ -33,6 +33,11 @@ public function getDescription();
      */
     public function getImage();
 
+    /**
+     * @return boolean
+     */
+    public function isImageAvailable();
```
Comment (kunicmarko20) [inline src/Meta/MetadataInterface.php]: bc break

Comment (jo66) [reply src/Meta/MetadataInterface.php]: Removed change on interface

**Comment by greg0ire** [comment]
I removed "add an upgrade note" from  your todo list, since there is no BC-break. Please fix the FlintCi build

**Comment by jo66** [comment]
@greg0ire Thanks, and I removed "update the docs", as it's nothing to document here

File: tests/Meta/MetadataTest.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace Sonata\BlockBundle\Tests\Meta;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\BlockBundle\Meta\Metadata;
+
+/**
+ * @author Jonathan Vautrin <jvautrin@pro-info.be>
+ *
+ * @group legacy
```
Comment (greg0ire) [inline tests/Meta/MetadataTest.php]: Why?

Comment (jo66) [reply tests/Meta/MetadataTest.php]: I even don't know, copy pasted, now removed.

File: tests/Meta/MetadataTest.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    public function testImageFalseDisableDefaultImage()
+    {
+        $metadata = new Metadata('title', 'description', false);
+        $this->assertFalse($metadata->getImage());
+    }
+
+    /**
+     * @dataProvider isImageAvailableProvider
+     */
+    public function testIsImageAvailable($title, $description, $image, $expected)
```
Comment (OskarStark) [inline tests/Meta/MetadataTest.php]:
```suggestion
    public function testIsImageAvailable($expected, $image)
```

title and description can be hardcoded and removed from the data provider, and please put expected as first, as phpunit assert* method use the expected value as the first argument

File: tests/Meta/MetadataTest.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        $metadata = new Metadata('title', 'description', $image);
+        $this->assertEquals($expected, $metadata->isImageAvailable());
```
Comment (OskarStark) [inline tests/Meta/MetadataTest.php]:
```suggestion
        $this->assertEquals(
            $expected,
            (new Metadata('title', 'description', $image))->isImageAvailable()
        );
```

File: src/Meta/Metadata.php [hunk 5]
```diff
@@ -78,11 +78,11 @@ public function getImage()
     }
 
     /**
-     * @return boolean
+     * @return bool
      */
     public function isImageAvailable()
     {
-        return $this->image !== false && $this->image !== self::DEFAULT_MOSAIC_BACKGROUND;
+        return false !== $this->image && self::DEFAULT_MOSAIC_BACKGROUND !== $this->image;
```
[the remark this answers is not in the document]
Comment (jo66) [reply src/Meta/Metadata.php]: That's what I was thinking, then I found this in test (from core bundle):
```
    public function testImageFalseDisableDefaultImage()
    {
        $metadata = new Metadata('title', 'description', false);
        $this->assertFalse($metadata->getImage());
    }
```
So I guessed that set to false is the way to disable background image...
But maybe it's only related to deprecated version from core bundle? In this case I should not manage false and remove the test above (that I copy pasted)?

[the remark this answers is not in the document]
Comment (core23) [reply src/Meta/Metadata.php]: Maybe you could fix this with an assert in the `constructor` method

[the remark this answers is not in the document]
Comment (jo66) [reply src/Meta/Metadata.php]: So to use a flag variable initialized in constructor? It was my first approach but @greg0ire suggested not to use variable and to compare againt value (see [review](https://github.com/sonata-project/SonataCoreBundle/pull/638#pullrequestreview-187665239)), but it was before we noticed that image can be null...

[the remark this answers is not in the document]
Comment (core23) [reply src/Meta/Metadata.php]: > So to use a flag variable initialized in constructor?

No, you could just use an assert in the constructor:

```php
   public function __construct($title, $description = null, $image = null, $domain = null, array $options = [])
    {
        if (!is_string($image) && null !== $image) {
            throw new InvalidArgumentException(...);
        }
    }
``

[the remark this answers is not in the document]
Comment (jo66) [reply src/Meta/Metadata.php]: Yes but in this case people who decided to set $image to false to have no background will get an error.
Are you sure we must do it?

[the remark this answers is not in the document]
Comment (core23) [reply src/Meta/Metadata.php]: As mentioned, this could be discussed in an other PR.

[the remark this answers is not in the document]
Comment (jo66) [reply src/Meta/Metadata.php]: OK, so I resolve

**Comment by jo66** [comment]
Not sure this PR still make sense as it's now the new Metadata class from Admin bundle that is used.

**Comment by jo66** [comment]
@core23 @OskarStark Can I delete branch?

**Comment by OskarStark** [comment]
Yes

**Pull request closed without merging**